fix the imaginary number printing

dipietrantonio opened the pull request:
#### References to other Issues or PRs

Fixes #13977
 
#### Brief description of what is fixed or changed

As pointed out in the issue https://github.com/sympy/sympy/issues/13977, before this fix the program generates wrong octave code when it comes to imaginary numbers, e.g.```sqrt(3)i```. Now it is set to make multiplication explicit, i.e., ```sqrt(3)*i```. It may not be the best way since i dont know this project, but it is the fastest that solved my personal issue.

dipietrantonio committed:
fix the imaginary number printing

As pointed out in the issue https://github.com/sympy/sympy/issues/13977, before this fix the program generates wrong octave code when it comes to imaginary numbers, e.g.```sqrt(3)i```. Now it is set to make multiplication explicit, i.e., sqrt(3)*i. It may not be the best way since i dont know this project, but it is the fastest that solved my personal issue.

normalhuman reviewed (changes requested):
Thanks for the contribution, but there are several issues here.

First, there are [tests](https://github.com/sympy/sympy/blob/master/sympy/printing/tests/test_octave.py#L112) that affirm the present behavior, so the PR fails them

Second, `i` could be assigned a value in Octave, so it's not reliable to use it as imaginary unit. `1i` or `4i` etc can be used for sure. This printer will normally use `1i` to represent the imaginary unit (this is done below in the code).   

Third, the purpose of the if-statement you changed is to have nicer representation for things like `3+4*I` so they print as `3+4i` instead of `3+4*1i`. This is a reasonable thing to do, it just mistakenly thinks it applies to `sqrt(3)*I`. I suggest replacing `expr.as_coeff_Mul()[0].is_integer` with `(S.ImaginaryUnit*expr).is_Integer`. This will surely restrict the contracted form to integer coefficients.

And a test should be added (see the link to tests above) for the issue being fixed:
```
assert mcode(sqrt(3)*I) == "sqrt(3)*1i"
```

cbm755 commented:
Thanks @Halolegend94 for this so far!  I agree with @normalhuman that we want `4i` and `sqrt(3)*1i`.

dipietrantonio committed:
Fixes imaginary number code generation.
Restricts nice complex number representation to literals.
Adds test to test if imaginary expressions are printed correctly.

Halolegend94 commented:
Thanks @normalhuman for the tips. I have implemented and committed your suggestions.

cbm755 commented:
Thanks both.  @normalhuman we'll need to add @Halolegend94 to the AUTHORS file; I dimly recall that procedure has changed but I've been a bit out of touch.

asmeurer commented:
Run `./bin/authors_update.py` and follow the instructions.

cbm755 commented:
thanks, I've done that
